bansa/src/main/kotlin/com/brianegan/bansa/reselect.kt

```diff
@@ -0,0 +1,235 @@
+            return computeFun(i[0] as I0, i[1] as I1, i[2] as I2, i[3] as I3)
+        }
+
+        override operator fun invoke(state: S): O {
+            return memoizer.memoize(
+                    si0(state),
+                    si1(state),
+                    si2(state),
+                    si3(state)
+            )
+        }
+    }}
+
+
+class SelectorForP3<S,I0:Any,I1:Any,I2:Any>(val si0:SelectorInput<S,I0>,
+                                            val si1:SelectorInput<S,I1>,
+                                            val si2:SelectorInput<S,I2>
+)
+{
+    fun<I3:Any> withField(fn: S.() -> I3) = SelectorForP4<S,I0,I1,I2,I3>(si0, si1, si2,InputField(fn))
+    fun<I3:Any> withSelector(si:SelectorInput<S,I3>) = SelectorForP4<S,I0,I1,I2,I3>(si0, si1, si2,si)
+    fun<O> compute(computeFun: (I0, I1,I2) -> O)  = object : AbstractSelector<S, O>() {
+        override val computeandcount = fun(i: Array<out Any>): O {
+            ++_recomputations
+            @Suppress("UNCHECKED_CAST")
+            return computeFun(i[0] as I0, i[1] as I1, i[2] as I2)
+        }
+
+        override operator fun invoke(state: S): O {
+            return memoizer.memoize(
+                    si0(state),
+                    si1(state),
+                    si2(state)
+            )
+        }
+    }
+}
+
+class SelectorForP2<S,I0:Any,I1:Any>(val si0:SelectorInput<S,I0>,
+                                     val si1:SelectorInput<S,I1>) {
+    fun<I2:Any> withField(fn: S.() -> I2) = SelectorForP3<S,I0,I1,I2>(si0, si1,InputField(fn))
+    fun<I2:Any> withSelector(si:SelectorInput<S,I2>) = SelectorForP3<S,I0,I1,I2>(si0, si1,si)
+    fun<O> compute(computeFun: (I0, I1) -> O)  = object : AbstractSelector<S, O>() {
+        override val computeandcount = fun(i: Array<out Any>): O {
+            ++_recomputations
+            @Suppress("UNCHECKED_CAST")
+            return computeFun(i[0] as I0, i[1] as I1)
+        }
+
+        override operator fun invoke(state: S): O {
+            return memoizer.memoize(
+                    si0(state),
+                    si1(state)
+            )
+        }
+    }
+}
+class SelectorForP1<S,I0:Any>(val si0:SelectorInput<S,I0>) {
+    fun<I1:Any> withField(fn: S.() -> I1) = SelectorForP2<S,I0,I1>(si0,InputField(fn))
+    fun<I1:Any> withSelector(si:SelectorInput<S,I1>) = SelectorForP2<S,I0,I1>(si0,si)
+    fun<O> compute(computeFun: (I0) -> O)  = object : AbstractSelector<S, O>() {
+        override val computeandcount = fun(i: Array<out Any>): O {
+            ++_recomputations
+            @Suppress("UNCHECKED_CAST")
+            return computeFun(i[0] as I0)
+        }
+
+        override operator fun invoke(state: S): O {
+            return memoizer.memoize(
+                    si0(state)
+            )
+        }
+    }
+}
+/**
+ * wrapper class for Selector factory methods [create], that basically is used only to capture
+ * type information for the state parameter
+ */
+class SelectorFor<S> {
```

Comment:
If you'd like to stick with builders, I'd recommend a better name than `SelectorFor`. I'd love to see `Selector.Builder` or `Selector.for` (where `for` would be a method). But it feels awkward to add a preposition to a class name.